### ISSUE TITLE
Crash using eraser

### ISSUE THREAD

**Issue by rolandlo** [xournalpp/xournalpp#3640]
**Affects versions :**
 - OS: Linux (Ubuntu 21.10)
 - observed on X11
 - Which version of libgtk do you use: 3.24.30
 - Version of Xournal++: 1.1.0+dev (from Dec. 1, 2021)
 - Installation method: PPA

**Describe the bug**
While annotating a PDF the app sometimes (randomly) crashes (with signal 11) when I use the eraser via stylus button. This looks like a regression, as I didn't experience that problem before.

**To Reproduce**
1. Write something onto a PDF
2. Erase using the stylus button
3. Keep doing that until the app crashes

**Expected behavior**
No crash

**Additional context**
Here are two errorlogs: [errorlogs.2021_12_06.zip](https://github.com/xournalpp/xournalpp/files/7659359/errorlogs.2021_12_06.zip)

**Comment by Febbe** [comment]
Are you already investigating?

**Comment by rolandlo** [comment]
No, but as soon as I find some time, I will investigate the issue further.

**Comment by rolandlo** [comment]
It also happens without pdf background and also without stylus button, i.e. when using the eraser from the toolbar. And it happens pretty quickly, only takes a couple of seconds until the crash happens.

**Comment by rolandlo** [comment]
The crash happens in one of the two lines (sometimes the first line, sometimes the second line):
https://github.com/xournalpp/xournalpp/blob/00991b4290198deebe0a8fb083c93eff783b6e70/src/core/model/eraser/ErasableStroke.cpp#L89-L90

It's probably a regression introduced in #3042.

Is this a problem with concurrent access in different threads?

**Comment by rolandlo** [comment]
The most direct way to reproduce the crash (consistently) is to draw a line (stroke with 2 points) and to try to erase it.

**Comment by Febbe** [comment]
The complete class is a mess, there is a possible memory leak; also, there are erasions in between iterations. The erase function may or may not advance an iterator...

**Comment by Febbe** [comment]
I can't debug this on windows, somehow I get this error:

```
[proc] Executing command: P:\Qt\Tools\CMake_64\bin\cmake.EXE --build c:/Users/Febbe/workspace/xournalpp/build --config Debug --target xournalpp -j 6 --
[build] [6/129   0% :: 0.045] Building RC object src/CMakeFiles/xournalpp.dir/win32/xpp.rc.res
[build] FAILED: src/CMakeFiles/xournalpp.dir/win32/xpp.rc.res 
[build] x86_64-w64-mingw32-windres -O coff -i C:/Users/Febbe/workspace/xournalpp/build/src/win32/xpp.rc -o src/CMakeFiles/xournalpp.dir/win32/xpp.rc.res
[build] CreateProcess failed: The system cannot find the file specified.
[build] ninja: fatal: ReadFile: Das Handle ist ung³ltig.
[build] 
[build] Build finished with exit code 1
```

**Comment by bhennion** [comment]
> Is this a problem with concurrent access in different threads?

I don't think so. In the end, this function is only called on a clone of the class' PartList.
The issue is most likey an iterator invalidation due to the `list.erase` used all over.

Changing the type `PartList` from `std::vector` to `std::list` fixes the crash. I'll make a PR ASAP

**Comment by rolandlo** [comment]
> I can't debug this on windows, somehow I get this error:
> 
> ```
> [proc] Executing command: P:\Qt\Tools\CMake_64\bin\cmake.EXE --build c:/Users/Febbe/workspace/xournalpp/build --config Debug --target xournalpp -j 6 --
> [build] [6/129   0% :: 0.045] Building RC object src/CMakeFiles/xournalpp.dir/win32/xpp.rc.res
> [build] FAILED: src/CMakeFiles/xournalpp.dir/win32/xpp.rc.res 
> [build] x86_64-w64-mingw32-windres -O coff -i C:/Users/Febbe/workspace/xournalpp/build/src/win32/xpp.rc -o src/CMakeFiles/xournalpp.dir/win32/xpp.rc.res
> [build] CreateProcess failed: The system cannot find the file specified.
> [build] ninja: fatal: ReadFile: Das Handle ist ung³ltig.
> [build] 
> [build] Build finished with exit code 1
> ```

I have no problem compiling Xournal++ via 
````
cmake -GNinja ..
cmake --build . --config Debug --target Debug -j 6
`````
in the MinGW 64 console under Windows. I haven't seen the error message you got on my system.

The crash when deleting lines also appears on Windows.

**Comment by bhennion** [comment]
> The issue is most likey an iterator invalidation due to the `list.erase` used all over.

Sorry, invalidation by the `list.emplace` in `erasePart`.

**Comment by bhennion** [comment]
See #3641 for a fix. The type `PartList` should never have been a `std::vector` to begin with: there are a lot of `erase` and `emplace` operations on this container.

**Comment by Febbe** [comment]
@bhennion 

> The type PartList should never have been a std::vector to begin with: there are a lot of erase and emplace operations on this container.

I disagree with that, with today's hardware, even with DDR5 (it is a bit better though). There is nearly no reason to use a normal linked list's for anything. Due to the fact, that memory allocations of a linked list are ordered randomly in the memory it is very likely, that every access will cause a cache miss. This drawback is so bad, that an insertion / deletion in a vector has to occupy (random assumption of) 200x more time than in a nearly empty vector. Since copying a complete cache line costs nearly the same like one 8 byte value, the vector must have a size of 200x8 to be slightly slower than a linked list on one insertion.

The factor can even be improved by using batch removals (e.g. std::erase_if / std::remove_if).

**Comment by Febbe** [comment]
https://www.youtube.com/watch?v=LFv7XwgsdLY

**Comment by bhennion** [comment]
@Febbe That's interesting! I didn't know that. Thanks!
Do you want a better fix than #3641 by just being careful about invalidated iterators? This is certainly possible.

IMO, we should just merge #3641 for now (a slow eraser is better than a crashy eraser). Then, this entire question becomes useless if and when we merge #3532.

**Comment by Febbe** [comment]
Yep that's fine, the STL-list with objects is still way faster than the `GList*` containing Pointers :D
